File: algorithms/[1905]统计子岛屿/solution.cpp
// topics = ["深度优先搜索", "广度优先搜索"]

#include <vector>
#include <array>
#include <queue>
#include <utility>

using namespace std;

class Solution
{
private:
    static constexpr array<array<int, 2>, 4> dirs = {{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}};

    void dfs(vector<vector<int>> &grid, int x, int y)
    {
        int m = grid.size();
        int n = grid[0].size();
        for (int d = 0; d < 4; d++)
        {
            int nx = x + dirs[d][0];
            int ny = y + dirs[d][1];
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] == 1)
            {
                grid[nx][ny] = 0; // 避免重复遍历
                dfs(grid, nx, ny);
            }
        }
    }

public:
    /**
     * BFS
     * time O(mn), space O(mn), m 和 n 为 grid 长宽
     */
    int countSubIslands(vector<vector<int>> &grid1, vector<vector<int>> &grid2)
    {
        int m = grid1.size();
        int n = grid1[0].size();
        queue<pair<int, int>> q;

        auto bfs = [&]()
        {
            bool flag = true;
            while (!q.empty())
            {
                auto [x, y] = q.front();
                q.pop();
                if (grid1[x][y] == 0)
                {
                    flag = false;
                }
                for (int d = 0; d < 4; d++)
                {
                    int nx = x + dirs[d][0];
                    int ny = y + dirs[d][1];
                    if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid2[nx][ny] == 1)
                    {
                        q.emplace(nx, ny);
                        grid2[nx][ny] = 0; // 避免重复遍历
                    }
                }
            }
            return flag;
        };

        int res = 0;
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (grid2[i][j] == 1)
                {
                    q.emplace(i, j);
                    grid2[i][j] = 0; // 避免重复遍历
                    res += bfs();
                }
            }
        }
        return res;
    }

    /**
     * DFS
     * time O(mn), space O(mn), m 和 n 为 grid 长宽
     */
    int countSubIslands2(vector<vector<int>> &grid1, vector<vector<int>> &grid2)
    {
        int m = grid1.size();
        int n = grid1[0].size();

        // 第一次遍历, 先将 grid2 中不属于 grid1 子岛的岛屿淹掉
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (grid1[i][j] == 0 && grid2[i][j] == 1)
                {
                    grid2[i][j] = 0; // 避免重复遍历
                    dfs(grid2, i, j);
                }
            }
        }

        // 第一次遍历, 统计 grid2 剩余子岛的数量
        int res = 0;
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (grid2[i][j] == 1)
                {
                    res++;
                    grid2[i][j] = 0; // 避免重复遍历
                    dfs(grid2, i, j);
                }
            }
        }
        return res;
    }
};

File: algorithms/[1905]统计子岛屿/solution.py
# topics = ["深度优先搜索", "广度优先搜索"]

from typing import List, Tuple, Deque
from collections import deque


class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        """
        BFS
        time O(mn), space O(mn), m 和 n 为 grid 长宽
        """
        m, n = len(grid1), len(grid1[0])
        direction = [[-1, 0], [0, 1], [1, 0], [0, -1]]
        q: Deque[Tuple[int, int]] = deque()

        def bfs() -> bool:
            flag = True
            while q:
                i, j = q.popleft()
                if grid1[i][j] == 0:
                    flag = False
                for cx, cy in direction:
                    nx, ny = i + cx, j + cy
                    if 0 <= nx < m and 0 <= ny < n and grid2[nx][ny] == 1:
                        q.append((nx, ny))
                        grid2[nx][ny] = 0  # 避免重复遍历

            return flag

        res = 0
        for i in range(m):
            for j in range(n):
                if grid2[i][j] == 1:
                    q.append((i, j))
                    grid2[i][j] = 0  # 避免重复遍历
                    res += bfs()

        return res
